### PR TITLE
test(devtools-example, devtools-extension): Remove tinylicious and fix devtools registration 

### PR DESCRIPTION
**Pull request by chentong7**
## Description

This PR is to remove tinylicious and fix devtool registration in example test-app and browser test-app. Testing Front-end using puppeteer is not support two threads server, that's why we are removing tinylicious and have our own fluid server starting with test-app.

Also, we enable logger to pass into example model loader. This change will allow devtool panel to monitor telemetry logs.

## Example runs in example test-app

https://github.com/microsoft/FluidFramework/assets/114451900/18a7e4ec-de7e-4088-a500-e0a4e1b07cfa

## Example runs in extension test-app

https://github.com/microsoft/FluidFramework/assets/114451900/6d41ba6b-1df4-4cd7-aa22-de680c222264

[AB#2272](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/2272)

### CODE REVIEW

File: packages/tools/devtools/devtools-example/src/FluidObject.ts [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+		],
+		{},
+	);
+
+	public static getFactory(): DataObjectFactory<CollaborativeDevtoolExampleAppObjects> {
+		return this.factory;
+	}
+
+	protected async initializingFirstTime(): Promise<void> {
+		// Create the SharedString and store the handle in our root SharedDirectory
```
Comment (alexvy86) [inline packages/tools/devtools/devtools-example/src/FluidObject.ts:96]:
```suggestion
		// Create the shared objects and store their handles in the root SharedDirectory
```

File: packages/tools/devtools/devtools-example/src/FluidObject.ts [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+		const matrixDimension = 2; // Height and Width
+		emojiMatrix.insertRows(0, matrixDimension);
+		emojiMatrix.insertCols(0, matrixDimension);
+		for (let row = 0; row < matrixDimension; row++) {
+			for (let col = 0; col < matrixDimension; col++) {
+				const emojiCell = SharedCell.create(this.runtime);
+				emojiMatrix.setCell(row, col, emojiCell.handle);
+			}
+		}
+		this.root.createSubDirectory(this.initialObjectsDirKey);
```
Comment (alexvy86) [inline packages/tools/devtools/devtools-example/src/FluidObject.ts:111]: Feels like this extra subDirectory is not necessary...? We're putting all the shared objects directly in the root, which seems fine. `getRootObject()` could directly return `this.root`... I think.

File: packages/tools/devtools/devtools-example/src/App.tsx [hunk 3]
```diff
@@ -242,37 +70,48 @@ function useContainerInfo(
...
+		async function getPrivateContainerData(): Promise<ContainerInfo> {
+			// Always create a new container for the private view.
+			// This isn't shared with other collaborators.
+			return createFluidContainer(loader);
 		}
 
 		getSharedFluidData().then((containerInfo) => {
 			if (getContainerIdFromLocation(window.location) !== containerInfo.containerId) {
 				window.location.hash = containerInfo.containerId;
+				document.title = containerInfo.containerId;
```
Comment (alexvy86) [inline packages/tools/devtools/devtools-example/src/App.tsx:86]: Maybe this?
```suggestion
				document.title = `Devtoolsl Example test app - ${containerInfo.containerId}`;
```

File: examples/utils/example-utils/src/modelLoader/modelLoader.ts [hunk 4]
```diff
@@ -56,14 +56,18 @@ export class ModelLoader<ModelType> implements IModelLoader<ModelType> {
 	// Here we specifically pick just the loader props we know we need to keep API exposure low.  Fine to add more
 	// here if we determine they're needed, but they should be picked explicitly (e.g. avoid "scope").
 	public constructor(
```
Comment (Josmithr) [inline examples/utils/example-utils/src/modelLoader/modelLoader.ts:58]: We should be sure to note in the PR description that we have added this capability to `example-utils`.

File: examples/utils/example-utils/src/modelLoader/modelLoader.ts [hunk 5]
```diff
@@ -56,14 +56,18 @@ export class ModelLoader<ModelType> implements IModelLoader<ModelType> {
...
-		props: Pick<ILoaderProps, "urlResolver" | "documentServiceFactory" | "codeLoader"> & {
+		props: Pick<
+			ILoaderProps,
+			"urlResolver" | "documentServiceFactory" | "codeLoader" | "logger"
```
Comment (Josmithr) [inline examples/utils/example-utils/src/modelLoader/modelLoader.ts:61]: We should be sure to note in the PR description that we have added the ability to provide a logger to `ModelLoader` in `example-utils`.

File: packages/tools/devtools/devtools-browser-extension/e2e-tests/chromebrowserdevtool.test.ts [hunk 6]
```diff
@@ -51,11 +51,11 @@ describe("End to end tests", () => {
...
 
 	// TODO
-	// it("Smoke: verify Devtools extension view can be launched", async () => {
-	// 	const targets = await browser.targets();
-	// 	console.log(targets);
-	// 	// chrome-extension://inmobceohkedafljagjfnbojplmlmgbk/devtools_app.html
-	// 	// document.querySelector("#containers-menu-section");
-	// 	// document.querySelector("#telemetry-menu-section");
-	// });
+	it("Smoke: verify Devtools extension view can be launched", async () => {
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-browser-extension/e2e-tests/chromebrowserdevtool.test.ts:54]: Should we revert these changes? I don't think they're related to this PR's goals.

File: packages/tools/devtools/devtools-example/package.json [hunk 7]
```diff
@@ -40,15 +39,18 @@
 	"dependencies": {
 		"@fluentui/react-components": "~9.19.1",
 		"@fluentui/react-icons": "^2.0.201",
+		"@fluid-example/example-utils": "workspace:~",
 		"@fluid-experimental/devtools": "workspace:~",
 		"@fluid-experimental/devtools-core": "workspace:~",
 		"@fluid-experimental/devtools-view": "workspace:~",
 		"@fluid-experimental/react-inputs": "workspace:~",
-		"@fluid-experimental/tree2": "workspace:~",
+		"@fluid-experimental/tree": "workspace:~",
```
[the remark this answers is not in the document]
Comment (Josmithr) [reply packages/tools/devtools/devtools-example/package.json:47]: Yeah, we don't want to depend on legacy shared-tree here. We must not actually be using `tree2` yet in the example, but that is what we will eventually use.

File: packages/tools/devtools/devtools-example/src/ClientUtilities.ts [hunk 8]
```diff
@@ -4,105 +4,104 @@
...
 
 	/**
 	 * The Container's unique ID. Also referred to as the "Document ID".
 	 */
 	containerId: string;
+
+	/**
+	 * Model object for the Container.
+	 */
+	obj: CollaborativeDevtoolExampleAppObjects;
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-example/src/ClientUtilities.ts:37]: Nit: I would probably call this `model` or `appModel`

File: packages/tools/devtools/devtools-example/src/ClientUtilities.ts [hunk 9]
```diff
@@ -4,105 +4,104 @@
...
 }
 
-function initializeTinyliciousClient(logger?: ITelemetryBaseLogger): TinyliciousClient {
-	console.log(`Initializing Tinylicious client...`);
-	return new TinyliciousClient({
+/**
+ * Creates a new Fluid Container from the provided client and container schema.
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-example/src/ClientUtilities.ts:41]: This comment is out of date.

File: packages/tools/devtools/devtools-example/src/ClientUtilities.ts [hunk 10]
```diff
@@ -4,105 +4,104 @@
...
+	const model = await loader.loadExisting(id);
+	const container = model.container;
+	return { container, containerId: id, obj: model.collaborativeObj };
 }
 
 /**
  * Creates a new Fluid Container from the provided client and container schema.
  *
- * @param containerSchema - Schema with which to create the container.
+ * @param loader - The Loader instance to use for creating the Container.
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-example/src/ClientUtilities.ts:84]: Nit
```suggestion
 * @param loader - The Loader to use for loading an existing Container or creating a new one.
```

File: packages/tools/devtools/devtools-example/src/ClientUtilities.ts [hunk 11]
```diff
@@ -4,105 +4,104 @@
...
+	return loader;
+}
+
+/**
+ * Creates a new Fluid Container from the provided client and container schema.
+ */
+async function createContainer(
+	loader: SessionStorageModelLoader<ICollaborativeDevtoolExampleAppModel>,
+): Promise<ContainerInfo> {
+	const createResponse = await loader.createDetached("1.0");
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-example/src/ClientUtilities.ts:63]: The old code did some console logging here to note each step in the creation / loading process. Could we re-add that to this new create/load code?

File: packages/tools/devtools/devtools-example/src/ClientUtilities.ts [hunk 12]
```diff
@@ -4,105 +4,104 @@
...
  *
- * @param containerSchema - Schema with which to create the container.
+ * @param loader - The Loader instance to use for creating the Container.
  * @param setContentsPreAttach - (optional) Callback for setting initial content state on the
  * container *before* it is attached.
- * @param logger - (optional) Telemetry logger to provide to client initialization.
  *
  * @throws If container creation or attaching fails for any reason.
  */
 export async function createFluidContainer(
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-example/src/ClientUtilities.ts:90]: I don't think we need these functions anymore. They basically just wrap calls to the new `createContainer` and `loadContainer` functions you added above.

File: packages/tools/devtools/devtools-example/src/Container.ts [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+import { IContainer } from "@fluidframework/container-definitions";
+import { requestFluidObject } from "@fluidframework/runtime-utils";
+
+import { CollaborativeDevtoolExampleAppObjects } from "./FluidObject";
+
+/**
+ * The CollaborativeDevtoolExampleAppModel is the model for our Fluid object.
+ * It contains the data we want to share with other clients.
+ */
+export interface ICollaborativeDevtoolExampleAppModel {
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-example/src/Container.ts:17]: Nit: We can probably just call this `AppModel`, since it's specific to the example app, and there is only 1 model.

File: packages/tools/devtools/devtools-example/src/Container.ts [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+
+import { ModelContainerRuntimeFactory } from "@fluid-example/example-utils";
+import { IContainerRuntime } from "@fluidframework/container-runtime-definitions";
+import { IContainer } from "@fluidframework/container-definitions";
+import { requestFluidObject } from "@fluidframework/runtime-utils";
+
+import { CollaborativeDevtoolExampleAppObjects } from "./FluidObject";
+
+/**
+ * The CollaborativeDevtoolExampleAppModel is the model for our Fluid object.
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-example/src/Container.ts:14]: Nit: I generally recommend not referring to the thing being documented by name in the comments. The comment is attached to the object, so calling out the name here isn't really necessary.

Comment (Josmithr) [reply packages/tools/devtools/devtools-example/src/Container.ts:14]: (same with some of the other type comments in this file)

File: packages/tools/devtools/devtools-example/src/Container.ts [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+	public constructor(
+		public readonly collaborativeObj: CollaborativeDevtoolExampleAppObjects,
+		public readonly container: IContainer,
+	) {}
+}
+
+const collaborativeObjId = "collaborative-obj";
+
+/**
+ * The CollaborativeTextContainerRuntimeFactory is the container code for our Fluid object.
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-example/src/Container.ts:32]: The name here is out of date, but I also don't think we need to name it.

E.g.
```suggestion
 * The runtime factory for the app.
```

File: packages/tools/devtools/devtools-example/src/Container.ts [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+ */
+export class CollaborativeDevtoolExampleObjectsContainerRuntimeFactory extends ModelContainerRuntimeFactory<ICollaborativeDevtoolExampleAppModel> {
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-example/src/Container.ts:34]: Since this is specific to our example app, and it's the only runtime factory, I think we can probably just call this "RuntimeFactory".

File: packages/tools/devtools/devtools-example/src/FluidObject.ts [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+import { SharedTree } from "@fluid-experimental/tree";
+import { SharedCounter } from "@fluidframework/counter";
+import { SharedString } from "@fluidframework/sequence";
+import { SharedCell } from "@fluidframework/cell";
+import { SharedMatrix } from "@fluidframework/matrix";
+
+/**
+ * CollaborativeDevtoolExampleAppObjects uses the React CollaborativeTextArea to load a collaborative HTML <textarea>
+ */
+export class CollaborativeDevtoolExampleAppObjects extends DataObject {
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-example/src/FluidObject.ts:16]: What do you think about just calling this something like "AppData"?

File: packages/tools/devtools/devtools-example/webpack.config.js [hunk 18]
```diff
@@ -8,12 +8,14 @@ const HtmlWebpackPlugin = require("html-webpack-plugin");
...
 
 const sourcePath = path.join(__dirname, "src");
+const webpack = require("webpack");
 
 module.exports = (env) => {
 	const isProduction = env && env.production;
 
 	return merge(
 		{
+			mode: isProduction ? "production" : "development",
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-example/webpack.config.js:18]: I think we can always use "development" here. We never intend to actually publish this example.
```suggestion
			mode: "development",
```

File: packages/tools/devtools/devtools-example/webpack.config.js [hunk 19]
```diff
@@ -25,6 +27,7 @@ module.exports = (env) => {
 					{
 						test: /\.tsx?$/,
 						loader: require.resolve("ts-loader"),
+						exclude: /node_modules/,
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-example/webpack.config.js:30]: I believe this is the default behavior. Do we need this line?

File: packages/tools/devtools/devtools-example/src/App.tsx (hunk 3 above)
Comment (chentong7) [reply packages/tools/devtools/devtools-example/src/App.tsx:86]: Updated! Thx!

File: packages/tools/devtools/devtools-example/src/FluidObject.ts (hunk 1 above)
Comment (chentong7) [reply packages/tools/devtools/devtools-example/src/FluidObject.ts:96]: Updated!

File: packages/tools/devtools/devtools-example/package.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (chentong7) [reply packages/tools/devtools/devtools-example/package.json:47]: Seems like the legacy shared tree can't add seriazed node in it. Move back to tree2

File: packages/tools/devtools/devtools-example/package.json [hunk 20]
```diff
@@ -40,15 +39,18 @@
 	"dependencies": {
 		"@fluentui/react-components": "~9.19.1",
 		"@fluentui/react-icons": "^2.0.201",
+		"@fluid-example/example-utils": "workspace:~",
 		"@fluid-experimental/devtools": "workspace:~",
```
[the remark this answers is not in the document]
Comment (chentong7) [reply packages/tools/devtools/devtools-example/package.json:43]: Right! Rmeove it!

File: packages/tools/devtools/devtools-example/src/FluidObject.ts (hunk 2 above)
Comment (chentong7) [reply packages/tools/devtools/devtools-example/src/FluidObject.ts:111]: Tested it. Actually we do need it or it would be "Error: InitialObjects sub-directory was not initialized".

File: packages/tools/devtools/devtools-example/src/FluidObject.ts (hunk 17 above)
Comment (chentong7) [reply packages/tools/devtools/devtools-example/src/FluidObject.ts:16]: Yeah. Sounds much better! Renamed it!

File: packages/tools/devtools/devtools-example/src/Container.ts (hunk 14 above)
Comment (chentong7) [reply packages/tools/devtools/devtools-example/src/Container.ts:14]: Fixed

File: packages/tools/devtools/devtools-example/src/Container.ts (hunk 13 above)
Comment (chentong7) [reply packages/tools/devtools/devtools-example/src/Container.ts:17]: Renamed

File: packages/tools/devtools/devtools-example/src/Container.ts (hunk 15 above)
Comment (chentong7) [reply packages/tools/devtools/devtools-example/src/Container.ts:32]: Fixed

File: packages/tools/devtools/devtools-example/src/Container.ts (hunk 16 above)
Comment (chentong7) [reply packages/tools/devtools/devtools-example/src/Container.ts:34]: Renamed

File: packages/tools/devtools/devtools-example/src/ClientUtilities.ts (hunk 11 above)
Comment (chentong7) [reply packages/tools/devtools/devtools-example/src/ClientUtilities.ts:63]: Adding logs for container creation and re-load.

File: packages/tools/devtools/devtools-example/src/ClientUtilities.ts (hunk 12 above)
Comment (chentong7) [reply packages/tools/devtools/devtools-example/src/ClientUtilities.ts:90]: Merge these functions

File: packages/tools/devtools/devtools-example/webpack.config.js (hunk 19 above)
Comment (chentong7) [reply packages/tools/devtools/devtools-example/webpack.config.js:30]: Tested it. Works without it.

File: packages/tools/devtools/devtools-browser-extension/e2e-tests/app/container.ts [hunk 21]
```diff
@@ -60,6 +60,7 @@ export class CollaborativeTextContainerRuntimeFactory extends ModelContainerRunt
  */
 class CollaborativeText extends DataObject {
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-browser-extension/e2e-tests/app/container.ts:61]: This isn't a collaborative text object, right? It's our entire app model?

File: packages/tools/devtools/devtools-browser-extension/e2e-tests/app/container.ts [hunk 22]
```diff
@@ -69,6 +70,10 @@ class CollaborativeText extends DataObject {
 		return this._text;
 	}
 
+	public get initialObjects(): Record<string, IFluidLoadable> {
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-browser-extension/e2e-tests/app/container.ts:73]: This isn't really a pattern we use with our 1P API. Instead of baking this in to our data object, I would recommend just passing in something like the following to the devtools initialization function:
```typescript
{ collaborativeText: text }
```

File: packages/tools/devtools/devtools-example/src/FluidObject.ts [hunk 23]
```diff
@@ -0,0 +1,270 @@
...
+
+	/**
+	 * Function to create an instance which contains getFactory method returning SharedTreeFactory.
+	 * The example application calls container.create() to create a new DDS, and the method requires:
+	 * #1. static factory method
+	 * #2. class object with a constructor returning a type with a handle field
+	 *
+	 * The function below satisfies the requirements to populate the SharedTree within the application.
+	 */
+
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-example/src/FluidObject.ts:197]: Nit
```suggestion
```

File: packages/tools/devtools/devtools-example/src/Container.ts [hunk 24]
```diff
@@ -0,0 +1,57 @@
...
+		super(
+			new Map([AppData.getFactory().registryEntry]), // registryEntries
+		);
+	}
+
+	/**
+	 * {@inheritDoc ModelContainerRuntimeFactory.containerInitializingFirstTime}
+	 */
+	protected async containerInitializingFirstTime(runtime: IContainerRuntime): Promise<void> {
+		const collaborativeText = await runtime.createDataStore(AppData.getFactory().type);
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-example/src/Container.ts:40]: Should we call this `dataStore` or something instead? Since it isn't a collaborative text element.

File: packages/tools/devtools/devtools-example/src/App.tsx [hunk 25]
```diff
@@ -401,65 +243,39 @@ function LoadingView(props: LoadingViewProps): React.ReactElement {
 /**
  * {@link AppView} input props.
  */
-interface AppViewProps extends ContainerInfo, HasContainerKey {}
+interface AppViewProps extends HasContainerKey {
+	/**
+	 * Container object.
+	 */
+	obj: AppData;
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-example/src/App.tsx:250]: Nit: How about `appData` instead of `obj`? It's a bit more descriptive.

**Review by Josmithr — APPROVED** [review]
This looks really great! Thank you for doing this work!

I left one pattern suggestion. The rest of my comments are nitpicks. I think after that, this is ready to go!

File: packages/tools/devtools/devtools-example/src/App.tsx (hunk 25 above)
Comment (chentong7) [reply packages/tools/devtools/devtools-example/src/App.tsx:250]: Renamed!

File: packages/tools/devtools/devtools-browser-extension/e2e-tests/app/container.ts (hunk 22 above)
Comment (chentong7) [reply packages/tools/devtools/devtools-browser-extension/e2e-tests/app/container.ts:73]: Yes. For the collaborative text input area, we can just pass the shared string, but we also want to show the container data in devtool `data` tag. This is the data object will be passed to register in devtool.

File: packages/tools/devtools/devtools-example/package.json [hunk 26]
```diff
@@ -57,10 +59,11 @@
 		"@fluidframework/map": "workspace:~",
 		"@fluidframework/matrix": "workspace:~",
 		"@fluidframework/protocol-definitions": "^1.1.0",
+		"@fluidframework/routerlicious-driver": "workspace:~",
```
Comment (alexvy86) [inline packages/tools/devtools/devtools-example/package.json:62]: One last nit from me: this package doesn't seem directly referenced, so wondering if it's actually required as a peer dep of something else or if we could remove it.

Comment (chentong7) [reply packages/tools/devtools/devtools-example/package.json]: Right. It can be removed. Thanks!

File: packages/tools/devtools/devtools-browser-extension/e2e-tests/app/container.ts (hunk 22 above)
Comment (Josmithr) [reply packages/tools/devtools/devtools-browser-extension/e2e-tests/app/container.ts:73]: I'm suggesting that we move the creation of this record to the site where we initialize the devtools. That would give us the same behavior, but would remove the concept of "initialObjects" from our data object, since it isn't really needed for anything else, and doesn't match our data object example conventions.

I don't think this is harmful as is, but I also don't think it's necessary.
